tempesta_fw/vhost.c

```diff
@@ -521,6 +555,176 @@ tfw_cleanup_locache(TfwCfgSpec *cs)
+
+	if (in4_pton(arg, -1, (u8 *)v4_saddr, delim, &pptr)) {
+		ipv6_addr_set_v4mapped(*v4_saddr, v6_inaddr);
+		tmpaddr.family = AF_INET;
+		goto done;
+	}
+	if (*arg == '[') {
+		arg++;
+		delim = ']';
+	}
+	if (!in6_pton(arg, -1, (u8 *)&v6_inaddr->s6_addr, delim, &pptr))
+		return NULL;
+	tmpaddr.family = AF_INET6;
+	if (*pptr == ']')
+		++pptr;
+done:
+	if (*pptr == '/')
+		++pptr;
+
+	*addr = tmpaddr;
+	return pptr;
+}
+
+/*
+ * Parse an IP address in CIDR format specified in the cache_purge_acl
+ * directive. That include an IPv4 or IPv6 address, and a potential
+ * address prefix specified as the number of bits after the '/' char.
+ * Returns zero if the result is positive.
+ * Returns a negative error number in case of an error.
+ */
+static int
+tfw_capuacl_parse_value(const char *arg, TfwAddr *addr)
+{
+	u8 prefix;
+	const char *pptr;
+
+	if ((pptr = tfw_capuacl_parse_addr(arg, addr)) == NULL)
+		return -EINVAL;
+
+	/* Parse a possible prefix length. */
+	if (*pptr == '\0') {
+		addr->in6_prefix = 128;
+		return 0;
+	}
+	if (kstrtou8(pptr, 10, &prefix) || (prefix == 0))
+		return -EINVAL;
+	if (addr->family == AF_INET) {
+		if (prefix > 32)
+			return -EINVAL;
+		prefix += 128 - 32;
+		addr->family = AF_INET6;
+	} else if (prefix > 128) {
+		return -EINVAL;
+	}
+	addr->in6_prefix = prefix;
+
+	return 0;
+}
+
+/*
+ * Process the cache_purge_acl directive.
+ */
+static int
+tfw_handle_cache_purge_acl(TfwCfgSpec *cs, TfwCfgEntry *ce)
+{
+	unsigned int i;
+	const char *val;
+	TfwVhost *vhost = &tfw_vhost_dflt;
+
+	if (ce->attr_n) {
+		TFW_ERR("%s: Arguments may not have the \'=\' sign\n",
+			cs->name);
+		return -EINVAL;
+	}
+	TFW_CFG_ENTRY_FOR_EACH_VAL(ce, i, val) {
+		TfwAddr addr = {};
+
+		if (tfw_capuacl_parse_value(val, &addr)) {
+			TFW_ERR("%s: Invalid ACL entry: '%s'\n",
+				cs->name, val);
+			return -EINVAL;
+		}
+		/* Make sure the address is not a duplicate. */
+		if (tfw_capuacl_lookup(vhost, &addr)) {
+			TFW_ERR("%s: Duplicate IP address or prefix: '%s'\n",
+				cs->name, val);
+			return -EINVAL;
+		}
+		/* Add new ACL entry. */
+		if (vhost->capuacl_sz == TFW_CAPUACL_ARRAY_SZ) {
+			TFW_ERR("%s: Unable to add new ACL: '%s'\n",
+				cs->name, val);
+			return -EINVAL;
+		}
+		vhost->capuacl[vhost->capuacl_sz++] = addr;
+	}
+	vhost->cache_purge_acl = 1;
+
+	return 0;
+}
+
+/*
+ * Process the cache_purge directive.
+ */
+static int
+tfw_handle_cache_purge(TfwCfgSpec *cs, TfwCfgEntry *ce)
+{
+	unsigned int i;
+	const char *val;
+	TfwVhost *vhost = &tfw_vhost_dflt;
+
+	if (ce->attr_n) {
+		TFW_ERR("%s: Arguments may not have the \'=\' sign\n",
+			cs->name);
+		return -EINVAL;
+	}
+	if (!ce->val_n) {
+		/* Default value for the cache_purge directive. */
+		vhost->cache_purge_mode = TFW_D_CACHE_PURGE_INVALIDATE;
+		goto done;
+	}
+	TFW_CFG_ENTRY_FOR_EACH_VAL(ce, i, val) {
+		if (!strcasecmp(val, "invalidate")) {
+			vhost->cache_purge_mode = TFW_D_CACHE_PURGE_INVALIDATE;
+		} else if (!strcasecmp(val, "purge")) {
```

Comment:
I think `cache_purge = purge` isn't clear enough and it'd be better to rename the option to `delete`, so one reading `cache_purge = delete` get more clear understanding what's going on.